### ISSUE TITLE
GTFS bounding box restriction isn't currently being applied

### ISSUE THREAD

**Issue by carlhiggs** [healthysustainablecities/global-indicators#281]
While investigating inconsistencies in GTFS feed reading by different packages in #275, it became apparent that during recent refactoring to address that issue the GTFS bounding box restriction isn't being applied.  For example, for an analysis for Bilbao, instead of restricting stops to the buffered urban region, they are included/analysed for all of Spain.

This isn't a problem for our main indicator analyses - ie. indicators for distance to closest destination will disregard inaccessible stops from beyond the buffered study region.  But it does mean that stops that are not relevant to a study region are being maintained and included in headway analysis, which increases processing time and data storage.  In addition the summary statistics displayed for the feeds are not specific to the study regions in question but to the configured feeds overall; the meaning of those is therefore not clear, when considered in terms of a specific study region.

This should be straightforward to address (ie. we already retrieve a bounding box for the buffered urban region; we just need to ensure that it is applied to restrict feeds).

This issue is impacting GTFS analysis (in the limited way described above; it doesn't impact the actual indicator estimates that would be calcualed) in the current `enhancements` and `gtfs` branches.

**Comment by carlhiggs** [comment]
So, when refactoring, I did make changes to the bounding box code on both branches, however the more suprising aspect to this is that the bounding box does not seem to be applied when using urbanaccess that explicitly at least intends to restrict to bbox (no such attempt is made with the code following reading by gtfslite):

https://github.com/global-healthy-liveable-cities/global-indicators/blob/a8178795cc4ec214ebcb93bb7f17c976ad35e998/process/subprocesses/_10_gtfs_analysis.py#L58-L78

While looking into this, I realised the loading of gtfs files using urbanaccess is being done through a helper script [ua_load.py](https://github.com/global-healthy-liveable-cities/global-indicators/blob/a8178795cc4ec214ebcb93bb7f17c976ad35e998/process/subprocesses/ua_load.py).  So its possible some of the issues encountered could be the result of aspects there --- although I can't see any obvious reason for the bounding box not being applied at first glance, 
https://github.com/global-healthy-liveable-cities/global-indicators/blob/a8178795cc4ec214ebcb93bb7f17c976ad35e998/process/subprocesses/ua_load.py#L291-L303

Looking at the urbanaccess.gtfs.utils_validation._validate_gtfs() function that is being called to remove stops, I think I now see the problem --- our bbox (changed during recent refactoring from geopandas to sql based query) is xmin, ymin, xmax, ymax; where as what is apparently required is (xmax, ymin, xmin, ymax):

https://github.com/UDST/urbanaccess/blob/9ccf3c92123fa884beb96878dc3465056c2bf323/urbanaccess/gtfs/utils_validation.py#L155

So, I made this change while writing this up -- with no apparent difference to the urbanaccess based results (still 771 stops, ie. Spain-wide rather than restricted to Bilbao for the Bilbao analysis).  

As per above -- this may not be an issue with urbanaccess per se, but how we're implementing it in the ua_load.py script.  But, it does seem to be that this is all over-complicated.

In principle, we could pass a bbox argument to the headway analysis function and restrict to relevant stops there regardless of package we use for reading.  If we implement this correctly, there really shouldn't be change in analysis results for indicators as the stops beyond the bounding box aren't accessible within the thresholds for analysis anyway.

So, what I think we could do is add the following after reading the feed using which ever package:
```
loaded_feeds.stops = loaded_feeds.stops.query(f"(stop_lat>={bbox['ymin']}) and (stop_lat<={bbox['ymax']}) and (stop_lon>={bbox['xmin']}) and (stop_lon<={bbox['xmax']})")
loaded_feeds.stop_times = loaded_feeds.stop_times.query(f"stop_id in {loaded_feeds.stops['stop_id'].values}")
```
For the RENFE AVLD feed in Bilbao, this restricts to 1 stop row and 24 stop_times rows.

Running the urbanaccess code with this modification on both gtfs and enhancements branches acts as intended, identifying the single rail stop from the RENFE AVLD feed that is located in Bilbao, both approximately looking like the following (with variation in analysis time)
```
Bilbao analysis:
  - /home/ghsci/work/process/data/transit_feeds/bilbao_gtfs/20230509_010334_RENFE_AVLD)
  - 20230508 to 20231231)
  - ['07:00:00', '19:00:00']


     Tram                  0 stops identified (0.27 seconds)
     Metro                 0 stops identified (0.26 seconds)
     Rail                  1 stops identified (0.31 seconds)
     Bus                   0 stops identified (0.25 seconds)
     Ferry                 0 stops identified (0.25 seconds)
     Cable tram            0 stops identified (0.25 seconds)
     Aerial lift           0 stops identified (0.25 seconds)
     Funicular             0 stops identified (0.26 seconds)
     Trolleybus            0 stops identified (0.25 seconds)
     Monorail              0 stops identified (0.25 seconds)

Bilbao summary (all feeds):
       tot_stops  headway<=30  headway<=20  pct_headway<=30  pct_headway<=20
mode                                                                        
Rail         1.0          NaN          NaN              NaN              NaN
total        1.0          0.0          0.0              0.0              0.0


pt_stops_headway exported to SQL database


Processing completed at 2023-05-19_1231
- Task: create study region boundary
- Duration: 0.05 minutes
```

Interactively querying details for this analysis, we see the day time weekday headway for RENFE AVLD rail services at this stop in Bilbao on average is about 199 minutes when analysed using gtfslite and 196 minutes when using urbanaccess, hence not meeting the 30 minute or 20 minute threshold criteria in either case.  As per #275 and #282 the difference is explained by urbanaccess (or our implementation of it) not using calendar_dates.txt, and hence overlooking the exceptions with some services cancelled resulting in less frequent service and higher headway interval estimates for this particular feed when read using gtfslite that takes these into account.

I have now committed changes made during the writing of this post, to both the `gtfs` and `enhancements` branch and so will close this issue.
